CX 21.1.0 libvkd3d shader

dasmy opened the pull request:
This PR is just to play around with building Vulkan, DXVK and VKD3D

dasmy committed:
play around with Vulkan and other build dependencies

Gcenx commented:
VKD3D requires the following to compile;
- flex
- MoltenVK
- pkgconfig 
- spirv-headers
- vulkan-headers
- widl

The _only_ reason to compile CX21 with VKD3D is to have access to WineD3D Vulkan backend for DirectX 10/11, but the performance is much slower than DXVK.

Gcenx commented:
For widl you can reuse the prebuilt binary from mingw-w64

Copy `$(brew —prefix mingw-w64)/toolchain-x86_64/bin/x86_64-w64-mingw32-widl` > `widl` making sure it's somewhere in $PATH this skips the need to compile winetools/widl on it's own.

What I do within my VKD3D Portfile;
```
pre-configure {
    file mkdir  ${workpath}/bin
    file copy   ${prefix}/bin/x86_64-w64-mingw32-widl ${workpath}/bin/widl
    configure.env-append    WIDL=${workpath}/bin/widl
    
    # Change the version number to our version number
    reinplace s|1.2|${version}|g ${worksrcpath}/configure.ac
}
```

dasmy commented:
> VKD3D requires the following to compile;
> 
> * bison (maybe 3.x)
> * flex
> * MoltenVK
> * pkgconfig
> * spirv-headers
> * vulkan-headers
> * widl
> 
> The _only_ reason to compile CX21 with VKD3D is to have access to WineD3D Vulkan backend for DirectX 10/11, but the performance is much slower than DXVK.

Thanks for the list.
As I am building MoltenVK (the one delivered in the Crossover-Tarball), lots of the stuff (spirv-headers, vulkan-headers) is already being downloaded automagically using `fetchDependencies` there.

dasmy commented:
> For widl you can reuse the prebuilt binary from mingw-w64
> 
> Copy `$(brew —prefix mingw-w64)/toolchain-x86_64/bin/x86_64-w64-mingw32-widl` > `widl` making sure it's somewhere in $PATH this skips the need to compile winetools/widl on it's own.
> 
> What I do within my VKD3D Portfile;
> 
> ```
> pre-configure {
>     file mkdir  ${workpath}/bin
>     file copy   ${prefix}/bin/x86_64-w64-mingw32-widl ${workpath}/bin/widl
>     configure.env-append    WIDL=${workpath}/bin/widl
>     
>     # Change the version number to our version number
>     reinplace s|1.2|${version}|g ${worksrcpath}/configure.ac
> }
> ```

This is a very very good hint - I already googled a bit and did not find anything apart from building the tools myself, which would be pretty annoying as building wine would have become a two-step process then... 👍

dasmy committed:
refine settings for vkd3d, see https://github.com/GabLeRoux/macos-crossover-wine-cloud-builder/pull/29#issuecomment-986839699

dasmy committed:
fix some errors I introduced

dasmy committed:
fix typos

Gcenx commented:
Ignoring those typos the real problem is crossover-sources DXVK is failing to compile.

dasmy commented:
> Ignoring those typos the real problem is crossover-sources DXVK is failing to compile.

Yes :-(

Looks like I will have to build locally to debug that problem...

Gcenx commented:
> > Ignoring those typos the real problem is crossover-sources DXVK is failing to compile.
> 
> Yes :-(
> 
> Looks like I will have to build locally to debug that problem...

Not really see https://github.com/doitsujin/dxvk/commit/f337ad3c05278e77a9500aa9d3d79b2b22ca2b89
Apply that on-top of the DXVK sources and we should be good to go for DXVK.

dasmy committed:
fix DXVK-build using https://github.com/doitsujin/dxvk/commit/f337ad3c05278e77a9500aa9d3d79b2b22ca2b89

Thx @Gcenx for pointing that out in https://github.com/GabLeRoux/macos-crossover-wine-cloud-builder/pull/29#issuecomment-986995068

Gcenx commented:
Canceled the actions as wine64 didn’t detect vkd3d, was this installed into /usr/local?, should add `--with-vkd3d` Into the wine64 stage for CX21+

Anything thing to consider is only building 64Bit DXVK dlls (also skipping d3d9 would be even more ideal)

dasmy commented:
Collecting all your comments here @Gcenx :

> I would recommend adding the following configure flag for VKD3D --without-ncurses stops opportunist linking (also handy for CX20/19)

Will do.

> Canceled the actions as wine64 didn’t detect vkd3d, was this installed into /usr/local?, should add `--with-vkd3d` Into the wine64 stage for CX21+

For now I just ran `make` but no `make install`. Somehow, I was hoping I could tell the wine `configure` where to find all the components instead of installing them to `/usr/local/...` and relying on everything being picked up properly.

> Another thing to consider is only building 64Bit DXVK dlls (also skipping d3d9 would be even more ideal)

I understand skipping 32bit DXVK, but why d3d9?
Also, is there a useful way to package DXVK properly with wine, so that I wouldn't have to download both separately and manually inject DXVK into my wine prefix (the Readme.md of DXVK suggests the latter)?

>

Another TODO: Should not forget to adapt `build_local.sh` in the end.

Gcenx commented:
> For now I just ran `make` but no `make install`. Somehow, I was hoping I could tell the wine `configure` where to find all the components instead of installing them to `/usr/local/...` and relying on everything being picked up properly.

While that's possible it doesn't make sense as the rpath would be saved into the compiled binary, unless the plan is to run install_name_tool after and fix the embedded rapth.
 
> I understand skipping 32bit DXVK, but why d3d9? 

MoltenVK doesn't support the required Vulkan APIs to enable the use of d3d9, this matches what CrossOver GUI actually does.

> Also, is there a useful way to package DXVK properly with wine, so that I wouldn't have to download both separately and manually inject DXVK into my wine prefix (the Readme.md of DXVK suggests the latter)?

The dlls could be shipped within the /lib64/wine/dxvk and place a modified version of the DXVK installer script into /bin.\
As for the idea to instead replace wine dlls with dxvk dlls won't work.

> Another TODO: Should not forget to adapt `build_local.sh` in the end.

Maybe for the moment it would be a good idea to also disabled all other builds and only allow CX21.1.0 and reenable the others once it's all complete.

dasmy committed:
try to ensure we pick up vkd3d properly

dasmy committed:
add --without-ncurses to VKD3D configure

dasmy commented:
> > For now I just ran `make` but no `make install`. Somehow, I was hoping I could tell the wine `configure` where to find all the components instead of installing them to `/usr/local/...` and relying on everything being picked up properly.
> 
> While that's possible it doesn't make sense as the rpath would be saved into the compiled binary, unless the plan is to run install_name_tool after and fix the embedded rapth.

Doesn't that mean, that if the wine configure were to pick up vkd3d from `/usr/local/...`, it would also have to reside there on a user's machine? - This would be rather inconvenient as this requires putting vkd3d into the proper place in addition to downloading, unpacking and running wine just from where it is.
(I am still hoping that we will eventually be able assemble everything required into a single tar-file, so that using wine-cx is almost just extract and run...).

> 
> > I understand skipping 32bit DXVK, but why d3d9?
> 
> MoltenVK doesn't support the required Vulkan APIs to enable the use of d3d9, this matches what CrossOver GUI actually does.

Disabling 32bit DXVK and/or d3d9 would require patching the build script. I am a bit hesitant doing this as every manually created patch will require some maintenance in future... :-)

> 
> > Also, is there a useful way to package DXVK properly with wine, so that I wouldn't have to download both separately and manually inject DXVK into my wine prefix (the Readme.md of DXVK suggests the latter)?
> 
> The dlls could be shipped within the /lib64/wine/dxvk and place a modified version of the DXVK installer script into /bin. As for the idea to instead replace wine dlls with dxvk dlls won't work.

Yes, this option sounds pretty convenient. - Will have a look.

> 
> > Another TODO: Should not forget to adapt `build_local.sh` in the end.
> 
> Maybe for the moment it would be a good idea to also disabled all other builds and only allow CX21.1.0 and reenable the others once it's all complete.

I am a bit afraid of breaking stuff for the older versions while adding things (btw.: shouldn't most of the stuff above also work for cx-20.x?). So I'd rather leave them on as long as GitHub does not complain about excessive usage of their resources.

However, it will possibly make sense to include some caching for the crossover source tar files to avoid looking like a DoS attack in their servers :-D

dasmy committed:
restore source tarballs from cache instead of downloading htem over and over again

Gcenx commented:
> Doesn't that mean, that if the wine configure were to pick up vkd3d from `/usr/local/...`, it would also have to reside there on a user's machine? - This would be rather inconvenient as this requires putting vkd3d into the proper place in addition to downloading, unpacking and running wine just from where it is. (I am still hoping that we will eventually be able assemble everything required into a single tar-file, so that using wine-cx is almost just extract and run...).

ld64 (Apples Linker) embedded the entire library path not just the name.\
See [Bugzilla 49199](https://bugs.winehq.org/show_bug.cgi?id=49199)

> Disabling 32bit DXVK and/or d3d9 would require patching the build script. I am a bit hesitant doing this as every manually created patch will require some maintenance in future... :-)

As the script lightly won't change much but this could be avoided by using `sed` to remove the offending line

> I am a bit afraid of breaking stuff for the older versions while adding things (btw.: shouldn't most of the stuff above also work for cx-20.x?). So I'd rather leave them on as long as GitHub does not complain about excessive usage of their resources.

I don't think it would break the other builds\
DXVK/MoltenVK should be fine for CX2*, VKD3D would be for CX21+\
Never tested DXVK on CX19 but we have been using MoltenVK-1.1.6 with DXVK from CX21 in CX20 without any major issues.

> However, it will possibly make sense to include some caching for the crossover source tar files to avoid looking like a DoS attack in their servers :-D

Agreed, I mirror the wine sources for a similar reason

Gcenx commented:
Done multiple edits to the last comment, mostly posting incase only an email notification was read.

dasmy committed:
set installation prefix for vkd3d directly inside wine installation (no adaptation of rpath, yet)

dasmy committed:
see if/how wine configure can deal properly with this

dasmy committed:
see if we get some better debug output

Gcenx committed:
build_monolithic.yml: some small fixes

- Separate wine build/runtime dependencies
- Install VKD3D into default location
- Install flex as wine build dep and add to $PATH
- Fix typo (22 to 20)
- Add note for install_name_tool place in action

VKD3D is better installed into the default location of /usr/local to make fixing the rpath simpler later and saves a headache of figuring out the correct PATH.

After this these can be copied into place to be bundled.

Gcenx committed:
configure_wine: cleanup options

"--disable-option-checking" makes configure will ignore invalid options not complain

Use "ac_cv_lib_soname_MoltenVK" to specify location instead of overriding LDFLAGS with an export

Gcenx committed:
build_monolithic.yml: more tweaks

Gcenx commented:
@dasmy this now builds again and VKD3D gets its files bundled within the wine installation.

Haven’t dealt with fixing rpath within the dylibs/*.so files but added sections to show where this should happen.

Building on Apple Silicon has been fixed in main so rebased the branch so there’s no conflicts.

dasmy committed:
see if we can properly add an rpath entry to automagically find vkd3d

dasmy commented:
@Gcenx , what is the reason you added two spots where `install_name_tool` should be called?

I now added a call to `install_name_tool` for inserting an `rpath` to `@executable/../lib` which should be sufficient (will verify after the build went through).

Also, shouldn't the vkd3d libs rather reside in `lib64` instead of `lib` ?

Gcenx commented:
@dasmy the first section is done before wine gets installed so only the bundled dylibs can be fixed (don’t want to touch libwine.dylib)

The second section is for fixing elf files (vkd3d-compiler binary etc)

Gcenx commented:
The reason vkd3d installed into /lib is that’s how the build was configured. It doesn’t really matter the location of the dylibs as long as the related elf can find it.

dasmy commented:
> @dasmy the first section is done before wine gets installed so only the bundled dylibs can be fixed (don’t want to touch libwine.dylib)
> 
> The second section is for fixing elf files (vkd3d-compiler binary etc)

I fixed the binaries where I felt, the `@rpath` was missing, with https://github.com/GabLeRoux/macos-crossover-wine-cloud-builder/pull/29/commits/1a9cb0cf6ce9477bb056fbb00f4579295b86f1ef . The vkd3d dylibs already had `@loader_path/.` in their `@rpath`, so I did not add any changes.

This way, the complaint of not finding `libvkd3d-shader` is gone. - Getting a black screen (with invisible UI elements that can be interacted with via blind clicking) with Quantum Odyssey now. 

Tess.exe (from https://github.com/KhronosGroup/MoltenVK/issues/628) complains that it `Cannot initialize the color shader object`.

Don't know if these errors are a local problem or related to any configuration issues.

Does it possibly make sense to bundle MoltenVK that we build in https://github.com/GabLeRoux/macos-crossover-wine-cloud-builder/blob/1a9cb0cf6ce9477bb056fbb00f4579295b86f1ef/.github/workflows/build_monolithic.yml#L183 with the package?

dasmy committed:
some update to the readme

Gcenx commented:
> I understand, this is just cosmetic. - Still a bit ugly, but surely not priority :-)

It takes _a lot_ of effort to make this clean, requires patching wine sources and manipulating the the make stage so not too practical.

> I fixed the binaries where I felt, the `@rpath` was missing, with [1a9cb0c](https://github.com/GabLeRoux/macos-crossover-wine-cloud-builder/commit/1a9cb0cf6ce9477bb056fbb00f4579295b86f1ef) . The vkd3d dylibs already had `@loader_path/.` in their `@rpath`, so I did not add any changes.
> 
> This way, the complaint of not finding `libvkd3d-shader` is gone. - Getting a black screen (with invisible UI elements that can be interacted with via blind clicking) with Quantum Odyssey now.

Ad long as it finds the needed library it’s all good, I left Macports to mostly deal with all that and later fix things as needed.

> Tess.exe (from [KhronosGroup/MoltenVK#628](https://github.com/KhronosGroup/MoltenVK/issues/628)) complains that it `Cannot initialize the color shader object`.
> 
> Don't know if these errors are a local problem or related to any configuration issues.

Not sure I’ve ran Witcher 3 using WineD3D Vulkan but went back to DXVK as it had less graphical glitches and was faster.

> Does it possibly make sense to bundle MoltenVK that we build in
> 
> https://github.com/GabLeRoux/macos-crossover-wine-cloud-builder/blob/1a9cb0cf6ce9477bb056fbb00f4579295b86f1ef/.github/workflows/build_monolithic.yml#L183
> 
> with the package?

If the intention is to make use of DXVK_macOS then the custom MoltenVK will be required.\
This isn’t the case if using WineD3D Vulkan.

dasmy committed:
do not build DXVK in 32bit

Gcenx commented:
I believe it’s best to move MoltenVK into its own action like llvm/clang.

llvm/clang and MoltenVK should always be compiled from the latest sources, possibly also do the same for DXVK?

dasmy commented:
> I believe it’s best to move MoltenVK into its own action like llvm/clang.
> 
> llvm/clang and MoltenVK should always be compiled from the latest sources, possibly also do the same for DXVK?

Do I understand correctly that you suggest to build MoltenVK (and possibly also DXVK) from 21.1.0 and then package/ship it also with the older versions (i.e. 20.x, 19.x)?

If yes:
* Can we guarantee that there are no incompatibilities to be expected?
* How would we properly package MoltenVK - just put `libMoltenVK.dylib` into `$wine-install/usr/local/lib` as you do with vkd3d?

Similar questions for DXVK:
* Is there a good spot inside the wine package where we can put the dlls and the installation script?

Gcenx commented:
> Do I understand correctly that you suggest to build MoltenVK (and possibly also DXVK) from 21.1.0 and then package/ship it also with the older versions (i.e. 20.x, 19.x)?
> 
> If yes:
> 
> * Can we guarantee that there are no incompatibilities to be expected?
> * How would we properly package MoltenVK - just put `libMoltenVK.dylib` into `$wine-install/usr/local/lib` as you do with vkd3d?

I’ve tested MoltenVK-v1.1.6 with wine-4.9 and managed to play VKQuake without issue so I don’t believe there’s any issues there though I’d skip DXVK for CX19.0.x as I’ve never tested on this only CX20.0.x

Yes place that into /lib, rpath can be fixed to look into this location.

> Similar questions for DXVK:
> 
> * Is there a good spot inside the wine package where we can put the dlls and the installation script?

I’d place the dlls into /lib64/wine/dxvk (like CrossOver)

Then modify the install script accordingly and place that into /bin making it self contained.

Gcenx commented:
Should also make copies like so;
- wine32on64 > wine
- wine32on64-preloader > wine-preloader

That would make this easier to use and also make it compatible with winetricks without needing to set environment exports.\
Making copies is better than using bash wrappers due to _SIP_ and symlinks always seemed to cause me more headaches for this.

dasmy commented:
Sorry, in the year-end chaos I did not find much time to continue here. Will pick this PR up again soon.

> Should also make copies like so;
> 
> - wine32on64 > wine
> 
> - wine32on64-preloader > wine-preloader
> 
> 
> 
> That would make this easier to use and also make it compatible with winetricks without needing to set environment exports.\
> 
> Making copies is better than using bash wrappers due to _SIP_ and symlinks always seemed to cause me more headaches for this.

That's a good idea (I do not see the issue about symlinks, though - do you have any details?).

Gcenx commented:
> Sorry, in the year-end chaos I did not find much time to continue here. Will pick this PR up again soon.

That’s life don’t worry about it.

> That's a good idea (I do not see the issue about symlinks, though - do you have any details?).

I can’t remember what the issues were just that making copies worked without issue.

dasmy committed:
create symlinks from wine to wine32on64 etc.

see https://github.com/GabLeRoux/macos-crossover-wine-cloud-builder/pull/29#issuecomment-995984096

dasmy commented:
> > Sorry, in the year-end chaos I did not find much time to continue here. Will pick this PR up again soon.
> 
> That’s life don’t worry about it.
> 
> > That's a good idea (I do not see the issue about symlinks, though - do you have any details?).
> 
> I can’t remember what the issues were just that making copies worked without issue.

Ok, I'll give it a try with symlinks as semantically, these seem clearer: When looking at them in the file system, it is obvious that these are avatars for the `32on64` executables. With copies, this is at least not obvious.

dasmy committed:
upload moltenvk artifact as preparation for factoring this out into a separate step

dasmy committed:
put MoltenVK into a proper place

dasmy committed:
include dxvk dlls in wine installation

dasmy committed:
always build MoltenVK and DXVK from latest sources (not integraated into wine, yet)

dasmy committed:
enable parallel build for MoltenVK dependencies

dasmy committed:
include spirv headers in package as VKD3D seems to need them for building

dasmy committed:
Remove parallel build parameter - does not seem to work properly

dasmy committed:
add missing mingw-w64 for DXVK

dasmy committed:
Merge branch 'cx-21.1.0_libvkd3d-shader' of https://github.com/GabLeRoux/macos-crossover-cloud-build into cx-21.1.0_libvkd3d-shader

dasmy committed:
fix typo

dasmy committed:
move moltenVK/DXVK build steps into separate action

dasmy committed:
make download and extract pre-created moltenVK/DXVK (and use the headers to build VKD3D)

dasmy committed:
some more on the DXVK install script

Gcenx commented on .github/actions/build_moltenvk_dxvk/action.yml:
```diff
@@ -55,7 +55,7 @@ runs:
 
         # disable building the 32bit version
-        sed -i "" "s/build_arch 32/# build_arch 32/g" ./package-release.sh
+        # sed -i "" "s/build_arch 32/# build_arch 32/g" ./package-release.sh
```
32Bit DXVK doesn’t work on macOS currently as wine(32on64) doesn’t have vulkan trunks.

Also d3d9 64Bit dlls can be skipped as those also don’t work due to lacking features of Metal/MoltenVK

dasmy committed:
add missing shell keyword in action

Gcenx commented:
@dasmy figure you’ve been busy, is there anything major that needs to still be taken into consideration for this?

dasmy commented:
> @dasmy figure you’ve been busy, is there anything major that needs to still be taken into consideration for this?

Correct: Work kept me a bit busy.
In principle, things should be in place. Proper removal of the 32bit stuff and D3DX9 has not been done, yet but this shouldn't hurt.

Unfortunately, I could not find any small Direct3D Windows executable that I could use for testing:
* dxdiag does not run because of some WinAPI function missing in wine
* a DirectX12 Tesselation Test that I found somewhere terminates with VKD3D as well as with DXVK complaining it is unable to create a proper surface. I could not find out if this is a hardware problem just on my computer (summer 2021 Intel-MacBookPro) or a more general issue, though.

Gcenx commented:
DirectX12 simply won’t function on macOS anyway due to Metal limitations, vkd3d is only there for the WineD3D Vulkan backend.

After I’m done with the Wine-7.0 packages I’ll cleanup this and squash merge.

Gcenx commented:
Keep getting side tracked so just squashed and merged while I can remember.
